server/game/cards/characters/01/melisandre.js

```diff
@@ -4,7 +4,8 @@ class Melisandre extends DrawCard {
     setupCardAbilities(ability) {
         this.reaction({
             when: {
-                onCardEntersPlay: (e, card) => card.controller === this.controller && card.hasTrait('R\'hllor')
+                onCardPlayed: (event, player, card) =>
```

Comment:
Switching to this event will break Mel for any instance where a R'hllor card is put into play, as this event is only fired when the player actively plays the card from hand.

In general I think this event should be moved to be fired *only* when an event is played, as it makes the semantics weird right now (implies that marshaling / ambushing is "playing" which does not match the language of the rules / cards.). Then you can have handlers for both events.